panda/src/cocoadisplay/cocoaGraphicsWindow.mm

```diff
@@ -269,11 +268,20 @@
  */
 void CocoaGraphicsWindow::
 end_flip() {
-  if (_gsg != (GraphicsStateGuardian *)NULL && _flip_ready) {
+  if (_gsg != (GraphicsStateGuardian *)NULL) {
 
     CocoaGraphicsStateGuardian *cocoagsg;
     DCAST_INTO_V(cocoagsg, _gsg);
-
+    
+    if (cocoagsg->will_vsync) {
+      if (cocoagsg->should_wait) {
+        cocoagsg->swap_lock.lock();
+        cocoagsg->swap_condition.wait();
+        cocoagsg->swap_lock.unlock();
+      }
+      cocoagsg->should_wait = true;
+    }
+    
```

Comment:
It's probably also slightly nicer to use `MutexHolder holder(gsg->_swap_lock);` than a lock/unlock pair in general (include `mutexHolder.h`), though not really a big deal.